File: memory/spm.py

import base
import cacti
import machine
import parser

def random_spm(machine, nxt, rand, cost):
   size = machine.word_size * 16
   spm = SPM(nxt, size)
   spm.reset(machine)
   while spm.get_cost() < cost:
      spm.size *= 2
      spm.reset(machine)
      if spm.get_cost() > cost:
         spm.size //= 2
         spm.reset(machine)
         break
   if spm.get_cost() <= cost:
      return spm
   return None

class SPM(base.Container):

   def __init__(self, mem, size = 0,
                access_time = 2,
                cycle_time = 2):
      base.Container.__init__(self, mem)
      self.size = size
      self.access_time = access_time
      self.cycle_time = cycle_time
      self.pending = 0

   def __str__(self):
      result  = "(spm "
      result += "(size " + str(self.size) + ")"
      result += "(access_time " + str(self.access_time) + ")"
      result += "(cycle_time " + str(self.cycle_time) + ")"
      result += "(memory " + str(self.mem) + ")"
      result += ")"
      return result

   def get_cost(self):
      if self.machine.target == machine.TargetType.SIMPLE:
         return self.size * 8
      elif self.machine.target == machine.TargetType.ASIC:
         return cacti.get_area(self.machine, self)
      elif self.machine.target == machine.TargetType.FPGA:
         width = self.machine.word_size * 8
         depth = self.size // self.machine.word_size
         return machine.get_bram_count(width, depth)
      else:
         assert(False)

   def permute(self, rand, max_cost):
      if self.size > self.machine.word_size and rand.randint(0, 1) == 1:
         self.size //= 2
      else:
         self.size *= 2
         if self.get_cost() > max_cost:
            self.size //= 2
            return False
      return True

   def simplify(self):
      self.mem = self.mem.simplify()
      if self.size == 0:
         return self.mem
      else:
         return self

   def reset(self, m):
      base.Container.reset(self, m)
      self.pending = 0
      if m.target == machine.TargetType.ASIC:
         self.access_time = cacti.get_access_time(m, self)
         self.cycle_time = cacti.get_cycle_time(m, self)

   def push_transform(self, index, rand):
      assert(index == -1)
      rand.push_limit(self.size, self.machine.addr_mask)

   def pop_transform(self, rand):
      rand.pop_limit()

   def get_path_length(self):
      return self.machine.addr_bits + self.get_next().get_path_length()

   def done(self):
      return max(self.pending - self.machine.time, 0)

   def process(self, start, write, addr, size):
      result = max(start, self.pending - self.machine.time)
      last_addr = (addr + size) & self.machine.addr_mask
      if addr < self.size and last_addr <= self.size:
         # Complete hits the scrachpad
         offset = addr % self.machine.word_size
         count = (size + self.machine.word_size + offset - 1)
         count //= self.machine.word_size
         self.pending = self.machine.time + result
         self.pending += max(self.cycle_time - self.access_time, 0)
         result += (count - 1) * self.cycle_time + self.access_time
      elif addr >= self.size and last_addr > self.size:
         # Completely misses the scratchpad
         self.pending = self.machine.time + result
         result = self.mem.process(result, write, addr, size)
      elif addr > self.size and last_addr < self.size:
         # First part hits, second part misses
         msize = size - last_addr + 1
         count = (last_addr + self.machine.word_size)
         count //= self.machine.word_size
         result += (count - 1) * self.cycle_time + self.access_time
         self.pending = self.machine.time + result
         self.pending += max(self.cycle_time - self.access_time, 0)
         result = self.mem.process(result, write, addr, msize)
      else:
         # First part misses, second part hits
         hsize = self.size - addr;
         offset = addr % self.machine.word_size
         count = (hsize + self.machine.word_size + offset - 1)
         count //= self.machine.word_size
         result += (count - 1) * self.cycle_time + self.access_time
         self.pending = self.machine.time + result
         self.pending += max(self.cycle_time - self.access_time, 0)
         result = self.mem.process(result, write, self.size, size - hsize)
      return result

def _create_spm(lexer, args):
   mem = parser.get_argument(lexer, args, 'memory')
   size = parser.get_argument(lexer, args, 'size', 0)
   access_time = parser.get_argument(lexer, args, 'access_time', 2)
   cycle_time = parser.get_argument(lexer, args, 'cycle_time', access_time)
   return SPM(mem, size, access_time, cycle_time)
base.constructors['spm'] = _create_spm

File: sim/spm.ml
open Base_memory

class spm =
    object (self)
        inherit container as super

        val mutable word_size : int = 0
        val mutable size : int = 0
        val mutable access_time : int = 0
        val mutable cycle_time : int = 0
        val mutable pending : int = 0

        method set name value = match name with
        | "word_size" -> word_size <- int_of_string value
        | "size" -> size <- int_of_string value
        | "access_time" -> access_time <- int_of_string value
        | "cycle_time" -> cycle_time <- int_of_string value
        | _ -> super#set name value

        method get_word_size = word_size

        method reset m main =
            super#reset m main;
            pending <- 0

        method finish = max (pending - mach#get_time) self#get_next#finish

        method private process_hit start write addr size =
            let offset = addr mod word_size in
            let count = (size + word_size + offset - 1) / word_size in
            pending <- mach#get_time + start;
            pending <- pending + max (cycle_time - access_time) 0;
            start + (count - 1) * cycle_time + access_time

        method private process_miss start write addr size =
            pending <- mach#get_time + start;
            send_request self#get_next start write addr size

        method private process_hit_miss start write addr size =
            let last_addr = (addr + size) land mach#get_addr_mask in
            let msize = size - last_addr + 1 in
            let count = (last_addr + word_size) / word_size in
            let result = start + (count - 1) * cycle_time + access_time in
            pending <- mach#get_time + result;
            pending <- pending + max (cycle_time - access_time) 0;
            send_request self#get_next result write addr msize

        method private process_miss_hit start write addr size =
            let hsize = size - addr in
            let offset = addr mod word_size in
            let count = (hsize + word_size + offset - 1) / word_size in
            let result = start + (count - 1) * cycle_time + access_time in
            pending <- mach#get_time + result;
            pending <- pending + max (cycle_time - access_time) 0;
            send_request self#get_next result write size (size - hsize)
 
        method process start write addr size =
            let result = max start (pending - mach#get_time) in
            let last_addr = (addr + size) land mach#get_addr_mask in
            if addr < size && last_addr <= size then
                (* Completely hits in the scratchpad. *)
                self#process_hit result write addr size
            else if addr >= size && last_addr > size then
                (* Completely misses the scratchpad. *)
                self#process_miss result write addr size
            else if addr > size && last_addr < size then
                (* First part hits, second part misses. *)
                self#process_hit_miss result write addr size
            else
                (* First part misses, second part hits. *)
                self#process_miss_hit result write addr size

    end
